### DISCUSSION TITLE
Include installer for language servers

### DISCUSSION THREAD

**Discussion by joegm** [helix-editor/helix#8245] [Ideas]
I got this idea from a [neovim plugin](https://github.com/williamboman/nvim-lsp-installer) that handles the installation of language servers for you.

I understand this would be tricky to implement, but would help in making Helix more accessible to people coming from IDEs and would be a lot comfier than tracking down all the LS's yourself.

## Suggestions
- Have a command like `$ hx --language install <language>` or something, similar to grammars.
- Have a popup when you open a new file type, letting you to install the LS for that language.
- Have a UI for managing servers (like the nvim plugin). Accessed by `:languages` for example.

I'm most in favour of the latter option, built-in UI's are something that I like a lot about Helix.

**Comment by claes** [comment]
An alternative approach could be to use language servers running in containers. Possible this would be easier to maintain. I found for example 
https://github.com/lspcontainers

**Comment by archseer** [comment]
There's a lot of different ways to install LSPs and these might even vary per platform. In some cases you install a distribution package, sometimes it's an npm package. Yet other times it's a binary you need to place in a system-specific location. This is doable when you have a scripting language available, but I'd like to avoid doing all of this in Rust.

Containers could be an option if there's a good set of containers available. It's still a bit tricky because you'd have to configure them to have access to the workspace files.

**Comment by domhel** [comment]
As someone who never understood LSP installation until yesterday, an installer would've been nice, but not necessary.
What I needed was a clear noob-friendly tutorial in the getting started section, that just tells me that I need LSP's in order to work with programming languages and that I just need to install them mostly with pip/npm and confirm their installation with `hx --health`.
If such a tutorial is within the getting started or even the installation process, I don't think it will be a problem.

**Comment by joegm** [comment]
> There's a lot of different ways to install LSPs and these might even vary per platform. In some cases you install a distribution package, sometimes it's an npm package. Yet other times it's a binary you need to place in a system-specific location. This is doable when you have a scripting language available, but I'd like to avoid doing all of this in Rust.

I understand that it would be more of a chore in rust and - like a lot of features - a plugin would probably be made when plugin support is complete. I just think that built in LSP features is one of the outstanding strong points of Helix and managing it from within the editor would contribute to that.

Obviously it's not necessary. Just having good docs on how to do it yourself is enough but it would be a cool thing to see. I'm not a good enough programmer to judge the scale of how tricky it would be, nevermind actually implementing it myself so feel free to close the issue if you're sure it wouldn't be worth it.

**Comment by rosvall** [comment]
I might be in the minority, but I **really** don't like it when my text editor suddenly starts downloading, compiling, and running tons of code from all over the internet. It's one of the things that really irks me about neovim's `lspconfig` plugin, and one of the reasons that I'm using helix now.

Edit: Maybe not lspconfig, but some plugin in my neovim setup _did_ do it.

**Reply by mati865** [reply]
It's `mason.nvim`.

**Comment by mjlbach** [comment]
Lspconfig does none of those things...

**Comment by KelvinJPS** [comment]
This wiki page https://github.com/helix-editor/helix/wiki/How-to-install-the-default-language-servers should be in the helix docs website. And the installer could just run npm commands?. Or it could be just providing information,  something like when you open a file, for example, of the extension C# it says. The language server for this language needs dotnet cli, and there is the omni-sharp website for more information.

**Comment by daedroza** [comment]
It is not that just language servers should be included. But also formatters and linters. The mason library on nvim does a great job for this.

**Reply by pascalkuthe** [reply]
As we already indicated we have no intention of turning helix into a package manager

**Reply by iampi31415** [reply]
Can helix use the binaries for LSPs and Linters, Formatters etc, directly from a path, so that I can do things like `"./helix/language.toml"` that points to those, and not to globals?

To put an example, you fork+clone project to work on, they use a set of lint+form+lsp. I need those to be used. Those are already in say "node_modules", I don't want to install them globally again (this will also bring up conflicting versions). Is this possible?

**Reply by CalebLarsen** [reply]
@iampi31415 Yes! That is possible. 
You can read more from the [documentation](https://docs.helix-editor.com/languages.html) but basically in your project root, create a folder called `.helix` and inside you can place your local `languages.toml`.

**Reply by iampi31415** [reply]
@CalebLarsen I've read through it but the "command" must always be in PATH so it's not possible to have it take the binary from node modules for each project for example, similar for python.

**Reply by sylvainterrien** [reply]
The command needs to be in PATH only if you just set the command name. You can also indicate the (absolute or relative) path to the binary, and it will run correctly even if not in PATH (just like in your shell).

An other solution, not specific to helix, would be to alter the PATH based on the current directory. A tool like [direnv](https://direnv.net/) does it very well.

**Reply by iampi31415** [reply]
I actually tried with full paths as well before commenting here, but got multiple issues (with pyright and ruff). I may repost at some point, currently I raised a similar query in Discussions. I think pyright  installed with `-g` may be necessary for some reason.
